File: step2/lec2/MergeSort.java
public class MergeSort {
    public static void main(String[] args) {
        int[] arr = new int[] { 9, 8, 1, 4, 5, 2, 3, 8, 10, 23, 4, 6, 8, 9, 0, 12, 1, 3, 45, 67 };

        mergeSort(arr, 0, arr.length - 1);

        print(arr);
    }

    static void print(int[] arr) {
        for (int i : arr) {
            System.out.print(i + " ");
        }

        System.err.println();
    }

    static void mergeSort(int[] arr, int start, int end) {

        if (start == end)
            return;

        int s = start;
        int middle = start + (end - start) / 2;

        mergeSort(arr, start, middle);

        mergeSort(arr, middle + 1, end);

        int[] temp = new int[end - start + 1];

        int m = middle + 1;

        int i = 0;

        while (start <= middle && m <= end) {
            if (arr[start] <= arr[m]) {
                temp[i++] = arr[start++];
            } else {
                temp[i++] = arr[m++];
            }
        }

        while (start <= middle) {
            temp[i++] = arr[start++];
        }

        while (m <= end) {
            temp[i++] = arr[m++];
        }
        System.err.println(start + ":" + end);
        print(temp);

        for (i = 0; i < temp.length; i++) {
            arr[s + i] = temp[i];
        }

    }

}

File: step2/lec2/mergesort.go
package main

import "fmt"

func main() {
	arr := []int{3, 2, 4, 0, -1, -5, 3}

	ms := &mergeSort{}

	arr = ms.Sort(arr, 0, len(arr)-1)

	fmt.Println(arr)

}

type mergeSort struct{}

func (m *mergeSort) Sort(arr []int, start int, end int) []int {
	if start == end {
		return arr
	}

	middle := start + (end-start)/2

	m.Sort(arr, start, middle)

	m.Sort(arr, middle+1, end)

	temp := make([]int, 0, end-start+1)

	i, j := start, middle+1

	for i <= middle && j <= end {
		if arr[i] <= arr[j] {
			temp = append(temp, arr[i])
			i++
		} else {
			temp = append(temp, arr[j])
			j++
		}
	}

	for i <= middle {
		temp = append(temp, arr[i])
		i++
	}

	for j <= end {
		temp = append(temp, arr[j])
		j++
	}

	for i = 0; i < len(temp); i++ {
		arr[start+i] = temp[i]
	}

	return arr
}
